Add betse.

leycec opened the pull request:
Greetings, tireless conda-forgers. This pull requests adds a simple recipe for [BETSE](https://gitlab.com/betse/betse), the pure-Python 3.x multiphysics biology simulator [my wife](https://www.researchgate.net/profile/Alexis_Pietak) and I actively maintain. A few ignorable caveats that no one probably cares about:

* Although pure-Python, I didn't mark this recipe as `noarch`. We require Python >= 3.4 for that sweet `Enum` action, which appears to contradict [brutal and unforgiving `noarch` constraints](https://conda-forge.org/docs/meta.html#building-noarch-packages). I'm surprised *any* packages satisfy those constraints, actually. We'd love to reduce your CI burden, but... I don't think we can. *Cue sad face.*
* Although we provide an extensive py.nose-based test suite of functional and unit tests, I decided *not* to exercise any of them in the `test:` section. I suspect you have better things to do with scarce CI resources than run expensive multiphysics simulations all day. Instead, I defaulted to just testing the importability of top-level subpackages and the executability of our entry point script. *Cue happy face.*
* I didn't know whether to shove build dependencies into the `requirements/build:` or `requirements/host:` lists. Since `conda skeleton pip` silently chose the latter *and* [official conda documentation](https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#host) also suggests the latter (e.g., "You should also include the base interpreter **[** in `host:`**]** for packages that need one. In other words, a Python package would list `python` here..."), I accepted this choice. That said, I have no idea what I'm doing. Hopefully, it's sufficient as is. *Cue confused face.*
* I didn't know how to avoid duplicating build dependencies between the `requirements/host:` and `requirements/run:` keys. Although the `build/run-exports:` key *appeared* to be of relevance, I found the [official conda documentation](https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#export-runtime-requirements) on that key to be unreadable. I briefly contemplated littering the top of the `meta.yaml` file with Jinja2 templating (e.g., `{% set python_req = "python >=3.4" %}`), but quickly realized that I was lazy. So, I just duplicated everything. *Cue lazy face.*
* I didn't know how to actually test whether my recipe does anything it claims to or not. I suspect that locally installing `conda-smithy` would help in some obscure manner, but... I failed to grep a working example. I'm probably just supposed to submit a pull request and let CI yell at me if it's broken. That's cool. I can do that.

That's about it. Thanks for the excellent community infrastructure, conda-forge. I can hardly fathom how much continued blood, sweat, and tears must go into this project. (But I bet it's *alot.*)

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/betse```) and found some lint.

Here's what I've got...


For **recipes/betse**:

 * Failed to even lint the recipe (might be a conda-smithy bug) :cry:

leycec commented:
`</sigh>`

*Aaaand...* we're off to a great start. Should I try firing off another linting attempt (e.g., by repushing or triggering the `@conda-forge-admin, please lint` webservice)? Well, let's try both! Presumably, I've done something horrifyingly awful in `meta.yaml`.

carlodri commented on recipes/betse/meta.yaml:
```diff
@@ -0,0 +1,86 @@
+
+package:
+  name: '{{ name|lower }}'
+  version: '{{ version }}'
```
remove single quotes

leycec commented:
@bgruening Thanks for that quick fix. Let's just try inlining the `hash_type` and `hash_value` Jinja2 variables. Why *does* `conda skeleton pip` templatize those, anyway? They're only ever referenced once. Such is code automation.

carlodri commented on recipes/betse/meta.yaml:
```diff
@@ -0,0 +1,86 @@
+  version: '{{ version }}'
+
+source:
+  fn: '{{ name }}-{{ version }}.{{ file_ext }}'
```
remove single quotes

leycec replied:
Ah, more `conda skeleton pip` inanity. I'll get this cleaned up post-haste. Thanks, @carlodri! You two are *really* on the ball here. :+1:

carlodri commented on recipes/betse/meta.yaml:
```diff
@@ -0,0 +1,86 @@
+source:
+  fn: '{{ name }}-{{ version }}.{{ file_ext }}'
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.{{ file_ext }}
+  {{ hash_type }}: '{{ hash_value }}'
```
remove single quotes

leycec replied:
On it! A *big* thanks, again.

dougalsutherland replied:
For the record @carlodri / @leycec, the single quotes here don't hurt anything.

leycec commented:
@conda-forge-admin, please lint

leycec commented:
I... don't know whether that actually does anything. My ignorance continues to shine brightly.

carlodri commented on recipes/betse/meta.yaml:
```diff
@@ -0,0 +1,84 @@
+source:
+  fn: {{ name }}-{{ version }}.{{ file_ext }}
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.{{ file_ext }}
+  sha256: 845896f4c30b3970d5bd9501e0ca754eef5af0d742de76103f66c8c8c58e4eca
```
set the `sha256` as a jinja variable at the top.

leycec replied:
Ah, right. One last edit for the evening, then. <sup>It's always "one last edit".</sup>

leycec commented:
Sadly, it's 5:08AM here in the frigid wastes of Ontario – which means I'm done for the day. If my `meta.yaml` is still broken (*it probably is*), I'll patch that up first thing in the morning.

Thanks for the astonishingly fast code review, @carlodri and @bgruening.

leycec commented:
*Urgh.* conda-forge-linter, y u make so crazy?

leycec commented:
Hmm; maybe it's the extraneous `file_ext` Jinja2 variable? I could try stripping that out, too. Or maybe I should just get some "beauty" sleep.

*Yeah.* Sleep is sounding pretty good.

leycec commented:
**Gotcha.** I *thought* that was probably the case, but sadly second-guessed myself. Here we go...

leycec commented:
*Woops.* The [CircleCI build failure](https://circleci.com/gh/conda-forge/staged-recipes/25596?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) is entirely my fault. I mistakenly assumed that, as Tkinter is typically installed with Python and PyQt5 is installed by default with Anaconda, that at least one would be the case here as well.

Thankfully, unit tests uncovered the shocking truth. Odd that Travis CI passed, but... *meh.* Circle CI is clearly in the right on this one.

I'll amend the `meta.yml` with an appropriate dependency. <sup>*crossing fingers, crossing fingers*</sup>

leycec commented:
**Yes!** Err, maybe yes. What say ye, @carlodri and/or @bgruening? Linting doesn't appear to be re-running, which mildly depresses me. Is there any way I can kick off a second conda-forge-linter run?

Thanks *yet again* for the immediate code review. I now sing the praises of the conda-forge team. :man_singer:

carlodri commented:
@leycec  think it's because it seems like you're amending the same commit over and over. Just make a new commit and the linter should trigger.

carlodri commented:
@conda-forge-admin, please lint

carlodri commented:
@conda-forge-admin, please lint

carlodri commented:
@conda-forge-admin, please rerender

leycec commented:
Right. Makes sense. Let's trigger this.

leycec commented:
@carlodri I may have accidentally broke your linter. :sob:

carlodri commented:
No changes, no linting... (I think)

leycec commented:
`</sigh>`

Quite right, quite right. Let's add a semicolon somewhere.

leycec commented:
Damn you, conda-forge-linter! No, seriously. I think your linter's busted. If it's not, I'll bust it myself.

leycec commented:
@carlodri Would it be feasible to forego conda-forge-linter and just visually lint the recipe by inspection? `meta.yaml` *is* only 85 lines, after all. The automation has [permanently fallen down and can't get up](https://www.youtube.com/watch?v=bQlpDiXPZHQ).

Here is where I play a sad violin.

carlodri commented:
To be honest, I have no idea of what's going on with the linter, the recipe seems syntactically correct to me. 

My guts say that you could remove the hosts section, which probably isn't necessary anyway

CJ-Wright commented on recipes/betse/meta.yaml:
```diff
@@ -0,0 +1,85 @@
+  entry_points:
+    - betse = betse.__main__:main
+  script: python -m pip install --no-deps --ignore-installed .
+  skip: True  # [py<34]
```
You don't need this line, and this can be noarch with a python version constraint.

leycec replied:
I *thought* that might be superfluous. I shall make it so – and, as you suggest, enable `noarch`.

A kitten for your troubles, Good Sir. :cat2:

CJ-Wright commented on recipes/betse/meta.yaml:
```diff
@@ -0,0 +1,85 @@
+requirements:
+  build:
+    - pip
+  host:
```
I don't know if we support host currently.

carlodri replied:
thanks for stepping in @CJ-Wright. Do you think this is why the linter fails?

carlodri replied:
judging from [this example](https://github.com/conda-forge/staged-recipes/pull/5581), seems like this is the reason. The linter should probably give a more descriptive error message.

leycec replied:
Well, that's a relief. I'm delighted the fix is trivial, but sorrowful that it was ultimately my fault. I blame `conda skeleton pip`. That's the first and last time I run *that* particular command.

Thanks to the power of ∞, @CJ-Wright and @carlodri.

CJ-Wright replied:
I don't know if I'd throw the too out completely, we just haven't moved to conda build 3 yet.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/betse```) and found it was in an excellent condition.

leycec commented:
oh. my. god.

conda-forge-linter *did* something.

leycec commented:
A **tl;dr** of changes in [the most recent commit](https://github.com/conda-forge/staged-recipes/pull/5574/commits/c1fe59a318532e85d9de10edbeae7c852186db66):

* Shifted all install-time requirements from the `host:` to `build:` section. 
* Dropped the `skip: True` line.
* Added a `noarch: python` line.

Linting and all CI hosts now pass, a beneficent miracle from on high.

leycec commented:
**Hrmm.** Spurious Windows-specific AppVeyor failure or actual bug? [You decide](https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.23070/job/i2f1peg37cgdu0j3):

```
Successfully built betse
Installing collected packages: betse
Successfully installed betse-0.7.0
INFO conda_build.build:bundle_conda(863): Packaging betse-0.7.0-0
number of files: 399
Fixing permissions
Traceback (most recent call last):
  File "C:\Miniconda-x64\Scripts\conda-build-all-script.py", line 5, in <module>
    exit(main())
  File "C:\Miniconda-x64\lib\site-packages\conda_build_all\cli.py", line 93, in main
    b.main()
  File "C:\Miniconda-x64\lib\site-packages\conda_build_all\builder.py", line 297, in main
    built_dist_location = self.build(meta, build_config)
  File "C:\Miniconda-x64\lib\site-packages\conda_build_all\builder.py", line 227, in build
    output_paths = conda_build.api.build(meta.meta, config=config)
  File "C:\Miniconda-x64\lib\site-packages\conda_build\api.py", line 97, in build
    need_source_download=need_source_download, config=config)
  File "C:\Miniconda-x64\lib\site-packages\conda_build\build.py", line 1524, in build_tree
    config=config)
  File "C:\Miniconda-x64\lib\site-packages\conda_build\build.py", line 1159, in build
    built_package = bundlers[output_dict.get('type', 'conda')](output_dict, m, config, env)
  File "C:\Miniconda-x64\lib\site-packages\conda_build\build.py", line 885, in bundle_conda
    files = post_process_files(metadata, initial_files)
  File "C:\Miniconda-x64\lib\site-packages\conda_build\build.py", line 852, in post_process_files
    noarch_python.populate_files(m, pkg_files, m.config.build_prefix, entry_point_script_names)
  File "C:\Miniconda-x64\lib\site-packages\conda_build\noarch_python.py", line 109, in populate_files
    os.unlink(src)
WindowsError: [Error 2] The system cannot find the file specified: u'C:\\bld\\betse_1523324596824\\_b_env\\Scripts\\betse-script.py'
Command exited with code 1
```

The [underlying BETSE project](https://gitlab.com/betse/betse) currently shows [tests passing with AppVeyor](https://ci.appveyor.com/project/betse/betse/branch/master). Is this ancient version of `conda-build` known to suffer race conditions against the notoriously fragile Windows filesystem... or something?

I'm sure this is my fault. Let's trigger another AppVeyor run with an empty commit to find out.

jakirkham commented on recipes/betse/meta.yaml:
```diff
@@ -0,0 +1,84 @@
+    - setuptools >=3.3
+    - six >=1.5.2
+    - ffmpeg
+    - pyside2
```
Would chop this list down to what is strictly needed for the build phase. So `python`, `pip`, `setuptools`. Guessing nothing else here is needed at build time.

leycec commented:
Naturally, I have broken something again. The combination of Windows + `noarch` + `entry_points` appears to be reliably failing for me on Windows. Specifically, the `noarch_python.populate_files()` function attempts to remove a non-existent entry point script.

I'm suffocatingly out of my depth here.

leycec commented:
@jakirkham Nice! Chopping the runtime requirements out of the build-time requirements list silently worked.

AppVeyor is still complaining vociferously. I'm pretty sure this is a `noarch` issue, as AppVeyor passed before switching to `noarch`. I'll remove that and try again.

jakirkham commented:
Yeah that usually helps with a variety of issues here and at the feedstock in addition to being a bit easier to maintain. :)

My guess is it is somehow getting hung up on the entry points. Wouldn't worry about it though as we only build `noarch` packages on Linux.

leycec commented:
> Yeah that usually helps with a variety of issues here and at the feedstock in addition to being a bit easier to maintain. :)

Preemptive problem solving: making everyone's life easier in the future, today. :rocket: 

Given how smoothly this went, I'll probably submit a followup recipe depending on this recipe at some point. Shout outs to everyone involved for the generous assistance.

> My guess is it is somehow getting hung up on the entry points.

*Yup.* Entry point scripts on Windows are bizarre and unfriendly.

I gouge my eyes out every time I re-examine the topic.

> Wouldn't worry about it...

How I have longed to hear these blessed words.

> ...as we only build `noarch` packages on Linux.

Makes sense! One final query: would you like me to clean up the commit history? Things got fairly hectic there. I'd be happy to rebase these six commits down into a single commit.

jakirkham commented:
> Given how smoothly this went, I'll probably submit a followup recipe depending on this recipe at some point.

Please do. If you have any thoughts on how we might improve the process, that would also be great to know.

> One final query, Commander: would you like me to clean up the commit history?

Wouldn't worry about that either. The history at staged-recipes is mainly for you. We only worry about the history for things like CI scripts and such in this repo. Recipes are transient in this repo (if everything is working correctly).

leycec commented:
> If you have any thoughts on how we might improve the process, that would also be great to know.

Why, I'm glad you asked. <sup>[insert cackling here]</sup>

Mostly, the process went great. The conda-build 3 issues are obviously transient, well-known, and probably about to go away. That said, there are three significant gotchas that snowblinded me. And I know *alot* about snow. More than I'd to know. It's literally snowing outside the cottage as I type this. Yes, I may be Canadian.

## Gotcha 1: conda-forge-linter

The conda-forge linter is clearly... a work in progress. That's understandable. Given scarce developer resources, I'd throw everything I had at the conda-build 3 migration too. Still, the linter sufficiently threw everyone for a loop here that it arguably squandered more developer time than it saved. That's bad. Automation is supposed to save time – not recklessly consume even *more* of it.

When resources permit,<sup>*...maybe after conda-build 3? maybe never?*</sup> it'd be awe-inspiring if you could throw a sacrificial developer or two at the following linter issues:

1. **The linter appears to silently fail after the first failure.** At the first failure, the linter helpfully emitted [an error message](https://github.com/conda-forge/staged-recipes/pull/5574#issuecomment-378863222). At each subsequent failure, the linter unhelpfully emitted... *nothing.* In fact, nobody even knew whether the linter was being re-run or not. Instead, I pompously propose the following:
    * Refactor the linter to **append one new comment to the current pull request after each (re-)linting of this request** (e.g., after a new commit triggers a github webhook triggering the linter). Even if that comment is the exact same as the prior comment that it appended for the prior linting, appending a new comment demonstrates to everyone involved that the linter is actually doing something. That's vital.
1. **The linter doesn't appear to emit meaningful errors.** At the first failure, the linter unhelpfully emitted [the error message](https://github.com/conda-forge/staged-recipes/pull/5574#issuecomment-378863222): "Failed to even lint the recipe (might be a conda-smithy bug) :cry: ". In actuality, the linter at least partially linted the recipe before fatally choking on the conda-build 3-specific `requirements/host:` list. Presumably, the linter internally raised an exception. But neither this exception nor any related error was conveyed to the user in the above error message. Instead, I pompously propose the following:
    * Refactor the linter to **embed exception messages and/or tracebacks in the comments that it appends to pull requests.** Exception messages would be equally meaningful for both conda-forge users and maintainers. Tracebacks would be utterly meaningless to users, of course, but probably invaluable to maintainers. In either case, something is better than nothing – and nothing is what we currently got.
1. **The linter appears to be overly draconian.** Like, [Soup Nazi draconian](https://www.youtube.com/watch?v=M2lfZg-apSA). On the one hand, it's a linter; it's *supposed* to be draconian. On the other hand, it appears to silently reject otherwise valid input that fails to conform to the exact example template for conda-forge `meta.yaml` files. Maybe? Nobody was quite sure about what was or was not permissible. Things were a bit... confusing. The prior two issues probably didn't help. I have no meaningful suggestions here, mostly because I'm unclear how strict (or not) the linter actually is. This rice ball adequately summarizes my confusion. :rice_ball: 
1. The `@conda-forge-admin, please lint` command didn't appear to do anything. I suspect this is because this command only applies to pull requests in existing feedstocks, whereas this is a pull request creating a new feedstock. This might also be a sub-issue of the first point above. In either case, it really didn't help.

In short, conda-forge-linter need work badly.

## Gotcha 2: conda skeleton pypi

[conda-forge documentation](https://conda-forge.org/docs/recipe.html#getting-started) on the `conda skeleton pypi {package_name}` command is scant at best:

> If it is a python package you can generate a skeleton as a starting point with `conda skeleton pypi your_package_name`. You do not have to use skeleton, and **the recipes produced by skeleton will need to be edited.**

That latter constraint should *definitely* be emphasized, ideally by linking to a new subsection of the existing [*Writing the `meta.yaml`*](https://conda-forge.org/docs/meta.html#writing-the-meta-yaml) section – titled, say, "Autogenerating the `meta.yaml` from PyPI." This subsection should explicitly list *all* of the changes that a user must manually make to the `meta.yaml` file autogenerated by the `conda skeleton pypi {package_name}` command. To my knowledge, this includes:

```
# Rewriting this...
{% set file_ext = "tar.gz" %}
{% set hash_type = "sha256" %}
{% set hash_value = "..." %}

package:
  name: '{{ name|lower }}'
  version: '{{ version }}'

source:
  fn: '{{ name }}-{{ version }}.{{ file_ext }}'
  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.{{ file_ext }}
  '{{ hash_type }}': '{{ hash_value }}'

# ...into this.
{% set sha256 = "..." %}

package:
  name: {{ name|lower }}
  version: {{ version }}

source:
  fn: {{ name }}-{{ version }}.tar.gz
  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
  sha256: {{ sha256 }}
```

```
# Rewriting this...
requirements:
   host:
     ...

# ...into this.
requirements:
   build:
     - pip
     ...
```

```
# Rewriting this...
build:
  ...
  script: python setup.py install --single-version-externally-managed --record=record.txt

# into this.
build:
  ...
  script: python -m pip install --no-deps --ignore-installed .
```

```
# Rewriting this...
about:
  license_file: ''

# ...into this, where {license_file} is the relative filename of this project's license relative to the project root.
about:
  license_file: {license_file}
```

```
# Rewriting this...
extra:
  recipe-maintainers: ''

# ...into this, where {github_username} is... your github username.
extra:
  recipe-maintainers:
    - {github_username}
```

Lastly, it might be helpful to note that all duplicate runtime requirements *except* Python and setuptools should typically be removed from the `requirements/build:` list: e.g.,

```
# Rewriting this...
requirements:
  build:
    - pip
    - python
    - setuptools >=3.3
    - numpy >=1.8.2
    - pillow >=2.3.0
    - scipy >=0.12.0
    - dill >=0.2.3
    - matplotlib >=1.5.0
    - six >=1.5.2
    - pyyaml >=3.10
  run:
    - python
    - setuptools >=3.3
    - numpy >=1.8.2
    - pillow >=2.3.0
    - scipy >=0.12.0
    - dill >=0.2.3
    - matplotlib >=1.5.0
    - six >=1.5.2
    - pyyaml >=3.10

# ...into this.
requirements:
  build:
    - pip
    - python
    - setuptools >=3.3
  run:
    - python
    - setuptools >=3.3
    - numpy >=1.8.2
    - pillow >=2.3.0
    - scipy >=0.12.0
    - dill >=0.2.3
    - matplotlib >=1.5.0
    - six >=1.5.2
    - pyyaml >=3.10
```

Most of this could (and probably should) be automated with a really hacky regex-based script of some sort runnable on user machines, which would post-process the `meta.yaml` file produced by `conda skeleton pypi` into a format consumable by conda-forge. Since nobody (*including me*) wants to write that script, publishing something resembling the above instructions might be the next best thing.

Let the user deal with it, I say!

## Gotcha 3: `noarch`

Modern conda-build and conda-forge support for the `noarch` flag is impressive. Tragically, [conda-forge documentation](https://conda-forge.org/docs/meta.html#building-noarch-packages) appears to be out-of-date with respect to this flag. Which is less impressive.

Specifically, I see two quandaries:

1. The condition "...and do not use selectors to skip building the recipe on a specific platform or for a specific version of python (e.g. `skip: True  # [py2k]`)." This is actually two-issues-in-one:
    1. The wording suggests the `noarch` flag applies *only* to pure-Python, cross-platform projects compatible with both Python 2.7 and Python 3.x. Instead, the `noarch` flag applies to all pure-Python, cross-platform projects that comply with the preceding conditions – regardless of their compatibility with Python 2.7 or Python 3.x.
    1. The `skip: True  # [py2k]` example implies that Python 3.x-only recipes should specify their requirement on Python 3.x with a selector. Instead, Python 3.x-only recipes should specify their requirement on Python 3.x with a minimum build- and runtime Python version requirement by:

```
# Rewriting this...
requirements:
  build:
    - pip
    - python
    ...
  run:
    - python

# ...into this, where {python_version} is the required Python 3.x version.
requirements:
  build:
    - pip
    - python >={python_version}
    ...
  run:
    - python >={python_version}
```

To clarify the above, it might be useful to add a new "Building `noarch` Python 3.x packages` subsection to the existing [*Writing the `meta.yaml`*](https://conda-forge.org/docs/meta.html#writing-the-meta-yaml) section. Python 3.x packages will probably comprise *most* new packages going forward, so it kinda seems sensible to target a discussion of how to `noarch` these packages.

This is probably the lowest priority of these three gotchas – but it's also probably the easiest to repair. Gotta love that sweet low-hanging fruit. :kiwi_fruit: 

## conda-forge: You're Still Awesome

All of the above is pedantic quibbling. conda-forge and its devoted development team are *fabulous* to work with. Everyone's done a smashing job. (I'm really impressed by this gargantuan scale of automation.)

conda-forge boggles the feeble mind. I'd be honoured to be a part of it.

CJ-Wright commented:
I think this is ready to go. Since it is noarch it doesn't need to build on anything other than circleci.

Thank you very much for such an in depth discussion of conda-forge issues, feedback is awesome!
Would it be possible to raise them as GitHub issues? I personally have an easier time when each thing is separated out into something I can close.

1. can go [here](https://github.com/conda-forge/conda-smithy/issues)
2. can go [here](https://github.com/conda/conda-build/issues) (although some of these issues are tied to conda-forge.
3. can go [here](https://github.com/conda-forge/conda-forge.github.io/issues)

Thank you very much!

CJ-Wright commented:
In the future please don't merge master to master, please make your own branch on your fork. Master to master merges require more scrutiny.

leycec commented:
> In the future please don't merge master to master, please make your own branch on your fork. Master to master merges require more scrutiny.

Absolutely. Apologies for the bureaucratic overhead. I'll shunt my next recipe submission into a  non-`master` branch, as you say. And thanks for those useful links, above; I'll get straight to the copying and pasting. :sun_with_face:

leycec commented:
**Bam!** At the cost of precious sleep, I have dutifully disaggregated my prior commentary into distinct issues. May they be trivial to address and plentiful in benefits.

jakirkham commented:
Thanks @leycec. The feedback is most welcome.

Also thanks for parsing it out on different issue trackers. That definitely helps stay on top of it.

leycec commented:
@jakirkham You're most welcome. I hope they haven't wasted *too* much of anyone's scarce time.

I've been pretty swamped for the past week-and-a-half myself, so I'm afraid I haven't been able to attend to this pull request. Did I need to trigger a @conda-forge-admin hook, shoot someone (or some*thing*) an e-mail, or otherwise directly involve myself in the eventual creation of a `betse` feedstock repository? I'd love to get the greasy ball rolling on that, but... I'm not sure how to do that.

Maybe it's time for a carcinogenic bottle of sake. :sake:

dougalsutherland commented:
@leycec It was actually created a while ago, and is available on conda: https://github.com/conda-forge/betse-feedstock

You should have gotten some automatic invites to the organization and to manage the repository. You're in charge now! The main unintuitive thing to remember is to [propose updates in PRs from a fork](https://conda-forge.org/docs/conda-forge_gotchas.html#using-a-fork-vs-a-branch-when-updating-a-recipe) rather than directly pushing to the repo or sending PRs from another branch on the main feedstock. Feel free to @- mention conda-forge/help-python or conda-forge/core if you run into issues in the future.

leycec commented:
Oh, super! Thanks for the judicious heads up.

> You should have gotten some automatic invites to the organization and to manage the repository.

I have good news and I have bad news.

* The good news is that BETSE now has a fully working Anaconda package. I've tested it locally. It behaves as expected. I can hardly believe it. Three mutated thumbs way up. :+1: :+1: :+1: 
* The bad news is that GitHub automation has failed us. I never received an automatic invite to anything. I've checked and rechecked my [GitHub dashboard](https://github.com), [notifications list](https://github.com/notifications?all=1), and [activity log](https://github.com/leycec). A whole lotta nuthin'. Presumably, GitHub hates me.

I'm not quite sure what happened here. I'm sure it's probably my fault, though. Would it be feasible to resend the invite? Alternately, would it be feasible to manually assign me as a repository developer of some sort? `</sigh>`

I know I've been an unmanageable pain. This probably means nothing, but I'm genuinely apologetic.

> The main unintuitive thing to remember is to propose updates in PRs from a fork...

**Right.** Automation concerns. That makes compelling sense. I'll absolutely do as you and the FAQ entry you helpfully linked to say. Also, consider giving whoever wrote that FAQ a raise. It's all really well-written.

Thanks again, @dougalsutherland. Sorry again.

isuruf commented:
![image](https://user-images.githubusercontent.com/5234427/39351118-189c291e-49c6-11e8-9bd4-14d886324cd5.png)

isuruf commented:
Try https://github.com/orgs/conda-forge/invitation

CJ-Wright commented:
@isuruf does that link work all the time? If so that would be great to put somewhere in the docs/readme so that new maintainers could just click on that to see what invitations they have.

isuruf commented:
Not sure. Found it in https://stackoverflow.com/a/44028010/4768820

jakirkham commented:
Indeed there is even an issue for that. ( https://github.com/conda-forge/conda-forge.github.io/issues/475 ) 😉

leycec commented:
Brilliant! I can confirm that [the link generously supplied by](https://github.com/orgs/conda-forge/invitation) @isuruf works exactly as advertised in a general-purpose manner. (Well, for me, at least.)

As suggested by @jakirkham at #475, it would probably be in everyone's best interests to visibly publish that link somewhere. A FAQ entry might be the short-term answer; an automated comment appended to the bottom of any newly accepted pull request might be the long-term answer.

I'm inclined to chalk this up to a GitHub API and/or UX issue of some unfathomable sort. I've been invited to numerous other GitHub organizations over the miserable span of my open-source "career." I can confirm that the conda-forge invitation is the only one to exhibit this aberrant behaviour. It simply didn't show up... *anywhere.* I didn't even receive a confirmatory e-mail at my linked Gmail account.

I doubt this issue is deterministically reproducible, else someone more intelligent than me should already have noted and resolved it. Presumably, it behaves as expected for most people. Sadly, I am not most people.

Thanks for all the conda-forge awesomeness.

leycec commented:
**EDIT:** *Yup.* According to [this pertinent StackOverflow thread](https://stackoverflow.com/a/44028010/4768820) (again supplied by @isuruf), this does appear to be a widespread GitHub API and/or UX issue. Organization invitations occasionally fail non-deterministically, for reasons no one knows. Some GitHubbers sporadically get hit by it. Some don't. I must have a bullseye painted on my chest. :dart: 

On the bright side, this isn't conda-forge's fault. On the dark side, you probably want to [publish that link](https://github.com/orgs/conda-forge/invitation).

jakirkham commented:
> ...an automated comment appended to the bottom of any newly accepted pull request might be the long-term answer

Thanks for the reminder. That has long been on the todo list. ( https://github.com/conda-forge/conda-forge-webservices/issues/73 ) We should really do that.

jakirkham commented:
Thanks for letting us know that link works.

Also sorry that GitHub gave you grief joining conda-forge. Agree generally we can document this better so that others won't have the same problem.

Most of us thinking about these things are far enough removed from being new users that we aren't always aware what a new user is looking for. So all of this feedback has been invaluable. Thanks for sharing with us. :)
